test/03_device-state.ts

```diff
@@ -510,6 +555,316 @@ export default () => {
+												api_heartbeat_state: (prop, value) =>
+													prop.to.equal(
+														DeviceOnlineStates.Online,
+														`The api_heartbeat_state should have recovered to online but it was found ${value}`,
+													),
+											},
+										);
+										expect(
+											await getHeartbeatWriteCacheState(device3.id),
+										).to.have.property('currentState', 'online');
+									},
+									/The api_heartbeat_state should have recovered to online but it was found offline/,
+								);
+
+								// TODO: This should not be happening. Drop this v test once the previous tests gets fixed.
+								itExpectsError(
+									'[FIXME] is expected that it will stay stuck to Offline when doing regular state GETs',
+									async function () {
+										await expectDeviceHeartbeat(device3.id, {
+											db: DeviceOnlineStates.Offline,
+											cache: DeviceOnlineStates.Online,
+										});
+										const keepPollingForMs =
+											devicePollInterval + TIMEOUT_SEC * 1000 + 1000;
+										const keepPollingForUntil = Date.now() + keepPollingForMs;
+										// Confirm that if the device keeps polling regularly every POLL_MSEC
+										// then it will stay as Offline in the DB.
+										while (Date.now() < keepPollingForUntil) {
+											await fakeDevice.getState(
+												device3,
+												device3.uuid,
+												stateVersion,
+											);
+											await setTimeout(maxGetStateEventConsumptionTimeout);
+											// B/c the write cache's state was already online, there was no device PATCH
+											expect(device3ChangeEventSpy.called).to.be.false;
+											await expectDeviceHeartbeat(device3.id, {
+												db: DeviceOnlineStates.Offline,
+												cache: DeviceOnlineStates.Online,
+											});
+
+											console.log(`⌚  Waiting for next state GET poll...`);
+											await setTimeout(POLL_MSEC);
+										}
+
+										expect(
+											await getHeartbeatWriteCacheState(device3.id),
+										).to.have.property('currentState', 'online');
+										await expectResourceToMatch(
+											pineUser,
+											'device',
+											device3.id,
+											{
+												// We use the same description as in the itExpectsError, so that we are sure which part failed.
+												api_heartbeat_state: (prop, value) =>
+													prop.to.equal(
+														DeviceOnlineStates.Online,
+														`The api_heartbeat_state should have recovered to online but it was found ${value}`,
+													),
+											},
+										);
+									},
+									/The api_heartbeat_state should have recovered to online but it was found offline/,
+								);
+							});
+
+							describe('when the write cache expires while the device is Online (eg API stops serving requests)', function () {
+								let device3: fakeDevice.Device;
+								const device3ChangeEventSpy = sinon.spy();
+
+								before(async function () {
+									device3 = await fakeDevice.provisionDevice(
```

Comment:
You also provision a device and wait for it to be Online in the above before. May be worth reducing code duplication and creating a shared function to do this.

Reply:
Actually it looks like two test cases are also quite similar. I first though we might could tidy things up by looping through `['Timeout', 'Online']` right above the describe level, but I guess the internal test logic is different enough that it wouldn't help.

Reply:
Yea, given the extra orchestration of setting up the spies & the stateMock, I think it's better to keep them separate, so that they also closely match the title in the describe, or it's more obvious if they do not match...